bc: fix for cross

dtzWill opened the pull request:
###### Motivation for this change

Significant rebuild, isolated change from #30882 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bgamari committed:
bc: fixes for cross-compilation

dtzWill commented:
Fixes `bc` cross, requires readline fix as well.

Tested with:

```
nix-build --arg crossSystem '(import ./lib).systems.examples.raspberryPi' -A bc
```

globin commented:
Please base this on staging :)

dtzWill commented:
@globin: oops, good call. Done.

globin commented:
cc @Ericson2314

dezgeg commented on pkgs/tools/misc/bc/default.nix:
```diff
@@ -7,11 +7,22 @@ stdenv.mkDerivation rec {
-  configureFlags = [ "--with-readline" ];
+  configureFlags = [
+    "--with-readline"
+    "CC_FOR_BUILD=${buildPackages.stdenv.cc.targetPrefix}gcc"
```
Should be done by stdenv and by default.

dtzWill replied:
Agreed, although actually I think we can omit this for now unless someone has a config where buildPackages' targetPrefix is not empty?

(pushed variant that drops this for now)

Ericson2314 reviewed (commented):
Hmm ideally we submit this patch somehow, and then pull it from bug tracker or whatever with `fetchpatch`.

dtzWill commented:
Readline fix submitted as #33576 , so as to not block on upstreaming the bc patch.

dtzWill commented:
Pushed with small fix for non-cross :).

Ericson2314 commented on pkgs/tools/misc/bc/default.nix:
```diff
@@ -9,9 +9,17 @@ stdenv.mkDerivation rec {
-  doCheck = true;
+  makeFlags = ''HOST_READLINE=${buildPackages.readline.out.outPath}/lib HOST_NCURSES=${buildPackages.ncurses.out.outPath}/lib'';
+
+  doCheck = stdenv.hostPlatform == stdenv.buildPlatform;
```
Can still be true now.

Ericson2314 commented on pkgs/tools/misc/bc/default.nix:
```diff
@@ -9,9 +9,17 @@ stdenv.mkDerivation rec {
+  patches = [ ./cross-bc.patch ];
+  nativeBuildInputs = [autoreconfHook flex ed.out texinfo] ++
+    stdenv.lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform)
+    [buildPackages.stdenv.cc buildPackages.readline.out buildPackages.ncurses.out];
```
BuildPackages.stdenv.cc should be a depsBuildBuild, can be unconditional too.

Ericson2314 replied:
The other two can be made unconditional too, based on the env vars below. Maybe add a comment saying for cross too since people might be confused without the conditional and remove it.

Ericson2314 reviewed (changes requested):
Thanks! Just a few things

Ericson2314 commented:
Weird, thanks for changing though! I'll take a look tomorrow or Monday. My only guess off hand is that turning the makeFlags into env var exports in preConfigure might help.

dtzWill commented:
rebased onto latest staging, dropped previous attempt to make use of `depsBuildBuild`.

I think this is finally good to go!

Builds for me natively and cross (tested with aarch64).

Ericson2314 commented:
@dtzWill I think we don't need the `HOST_` env vars anymore I'll check.

dtzWill commented:
@Ericson2314 in the patch, you mean? I already dropped them from the expression FWIW.

dtzWill commented:
@Ericson2314 guess don't need them in either place, pushed commit removing from patch.

Ericson2314 commented:
@dtzWill last thing is the native libraries really should be `nativeBuildInputs` not `depsBuildBuild`. Weird that this works. I'm going to investigate.

Ericson2314 commented:
OK, fixed the deps, and made the patch a bit more readable than what the diff algorithm did. Thanks!

dtzWill commented:
Yay! Thanks!
